src/main/java/httpserver/HttpRouter.java

```diff
@@ -0,0 +1,82 @@
+package httpserver;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An HttpRouter is used to route incoming requests to specific handlers.
+ *
+ * @see HttpHandler
+ * @see HttpRequest
+ */
+public class HttpRouter {
```

Comment:
Thought: remove routers entirely, and switch to using Handlers for everything (and rename handlers to routers).

(Every instance of "Router" onward means what is currently called `HttpHandler`)

New routers would have path routing stuff built in (as they currently do), and have the added ability of adding a "sub router", which would work the same way `addHandler` works in this class.

Basically you'd have code that looks something like this:

``` java
HttpServer s = new HttpServer();
s.get("/", new Route() { 
    @Override public void handle(HttpRequest req, HttpResponse resp) { 
        resp.write("hi."); 
    }
});

HttpRouter subRouter = new HttpRouter();
subRouter.get("/", new Route() {
    @Override public void handle(HttpRequest req, HttpResponse resp) {
        resp.setBody("sub router!");
    }
});

s.addRouter("/sub", subRouter);

s.run();
```

The only two paths that would match would be:

```
"/" => "hi."
"/sub/" => "sub router!"
```

Internally, the `Router#handle` method would first search to see if there's a subrouter for the first path segment, and try to use that. If there isn't, or if the remaining path segment doesn't match any of that router's routes, it'll go to the main routing mechanism currently in place. 

Hopefully that makes sense, it looks a little convoluted to me, and I'm the one trying to explain it.